core/src/main/scala-3/pureconfig/generic/derivation/CoproductConfigReaderDerivation.scala

```diff
@@ -0,0 +1,56 @@
+package pureconfig
+package generic
+package derivation
+
+import scala.compiletime.{constValue, erasedValue, summonFrom, summonInline}
+import scala.deriving.Mirror
+
+import pureconfig.error.{CannotConvert, ConfigReaderFailures}
+import pureconfig.generic.derivation.ConfigReaderDerivation
+import pureconfig.generic.derivation.WidenType.widen
+
+trait CoproductConfigReaderDerivation(fieldMapping: ConfigFieldMapping, optionField: String) {
+  self: ConfigReaderDerivation =>
+  inline def derivedSum[A](using m: Mirror.SumOf[A]): ConfigReader[A] =
+    new ConfigReader[A] {
+      def from(cur: ConfigCursor): ConfigReader.Result[A] =
+        for {
+          objCur <- cur.asObjectCursor
+          optCur <- objCur.atKey(optionField)
+          option <- optCur.asString
+          result <-
+            readers.get(option) match {
+              case Some(reader) => reader.from(cur)
+              case None =>
+                Left(
+                  ConfigReaderFailures(
+                    optCur.failureFor(
+                      CannotConvert(option, constValue[m.MirroredLabel], "The value is not a valid option.")
```

Comment:
I understand that we don't want to use `NoValidCoproductOptionFound` as it's defined in `generic-base`, but it's a bit disappointing that we are losing the rich message provided by it. Should we consider copying it to `derivation` or restructure that one in particular? What do you think, @jcazevedo?

Reply:
Some generic derivation is now part of core if we merge this, so it might make sense to have some basic error types for derivation and such also defined in core.

Reply:
An alternative is also to move all the new code into `generic-base`.

Reply:
I think `generic-base` is not an adequate place for actual derivation code. That module was created precisely to share common definitions across all derivation libraries. If we were to take this to another module I'd rather have a new `pureconfig-generic-scala3` that depended on `generic-base`, making it on par with shapeless and magnolia. However, given that this uses only Scala 3 native concepts I'm leaning towards keeping this in code in `core` and replicating generic classes to `core` in order to keep Scala 3 derivation self-contained.

Reply:
@ruippeixotog: That looks like a sensible way to go about it.